tests/lib/Service/Storage/DataSourceServiceTest.php

```diff
@@ -0,0 +1,248 @@
+
+        $dataSourceService = new DataSourceService(
+            [
+                new InMemoryDataSource($this->createItems()),
+            ],
+            $itemGroupStrategy
+        );
+
+        self::assertEquals(
+            $expectedGroupList,
+            $dataSourceService->getGroupedItems(
+                $criteria,
+                SupportedGroupItemStrategy::GROUP_BY_ITEM_TYPE_AND_LANGUAGE
+            )
+        );
+    }
+
+    /**
+     * @return iterable<array{
+     *  CriteriaInterface, ItemListInterface
+     * }>
+     */
+    public function providerForTestGetItems(): iterable
+    {
+        yield [
+            $this->itemCreator->createTestCriteria(
+                [],
+                ['pl']
+            ),
+            $this->itemCreator->createTestItemList(),
+        ];
+
+        yield [
+            $this->itemCreator->createTestCriteria(
+                [
+                    ItemType::ARTICLE_IDENTIFIER,
+                    ItemType::PRODUCT_IDENTIFIER,
+                ],
+                ['en']
+            ),
+            $this->itemCreator->createTestItemListForEnglishArticlesAndProducts(),
+        ];
+
+        yield [
+            $this->itemCreator->createTestCriteria(
+                [
+                    ItemType::ARTICLE_IDENTIFIER,
+                    ItemType::BLOG_IDENTIFIER,
+                    ItemType::PRODUCT_IDENTIFIER,
+                ],
+                ['en', 'fr', 'de', 'no']
+            ),
+            $this->createItems(),
+        ];
+    }
+
+    private function createDataSourceMockForCriteria(
+        CriteriaInterface $criteria,
+        ItemListInterface $expectedItemList,
+        int $expectedCount
+    ): DataSourceInterface {
+        /** @var DataSourceInterface|\PHPUnit\Framework\MockObject\MockObject $source */
+        $source = $this->createDataSourceMock();
+        $source
+            ->expects(self::once())
+            ->method('countItems')
+            ->with($criteria)
+            ->willReturn(
+                $expectedCount
+            );
+
+        if ($expectedCount > 0) {
+            $source
+                ->expects(self::once())
+                ->method('fetchItems')
+                ->with($criteria)
+                ->willReturn(
+                    $expectedItemList
+                );
+        }
+
+        return $source;
+    }
+
+    private function createDataSourceMockForGetItem(
+        string $id,
+        string $language,
+        ItemInterface $expectedItem
+    ): DataSourceInterface {
+        /** @var DataSourceInterface|\PHPUnit\Framework\MockObject\MockObject $source */
+        $source = $this->createDataSourceMock();
+        $source
+            ->expects(self::once())
+            ->method('fetchItem')
+            ->with($id, $language)
+            ->willReturn($expectedItem);
+
+        return $source;
+    }
+
+    private function createDataSourceMock(): DataSourceInterface
```

Comment:
In each instance this method is used, it's always also configured.
Either annotate method with union return type and drop the `@var` in calling code, or drop the method entirely (as it's just a simple mock creation, nothing more).